contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +900,34 @@ clearly described and the permissions issues should be resolved.
 This could be done with a double opt-in (an opt-in from both the referrer and the refer-ee) or with secondary permissions
 checks performed in admission. 
 
-TODO: Plugins, extensions, nested kinds, headers
+TODO: Plugins, extensions, headers
 
+### Handling references to multiple kinds
+
+References which can refer to multiple kinds should use a single field, and select the target kind via `apiVersion` and `kind` fields.
```

Comment:
For generic references, at minimum, you should include `group` and either `kind` or `resource`.

Are resolvers expected to use discovery information to resolve these references? If not, you need the following:
* resource (e.g. "deployments", not "Deployment")
* group
* version

If they _are_ expected to use discovery, you still need `group` and `kind`-or-`resource`, but can decide whether to include version or not.

Pinning a reference to a specific version has pros and cons:
* If the reference is pointing to a specific field path inside the object, that can only be assumed to be stable for a particular API version. If the reference is only used to locate the object or pull info out of the object's metadata, you don't actually need to pin to a particular version.
* Pro: Including the API version in the reference (along with group and resource) means resolvers don't have to use discovery information
* Con: Including the API version in the reference means that if the referenced API graduates and stops serving the old version, references go bad

cc @deads2k

Reply:
> For generic references, at minimum, you should include group and either kind or resource.

personally I prefer kind + group as the qualifiers to filter on, but [the existing conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references) recommend [ObjectReference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#objectreference-v1-core), which uses "apiVersion" instead.

> Are resolvers expected to use discovery information to resolve these references? If not, you need the following:

Today I believe that's largely based on the libraries available to help resolve references and retrieve objects. e.g. [controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#example-Client-Get).

[Controller-runtime looks to find the resource via using the group and version](https://github.com/kubernetes/client-go/blob/master/rest/url_utils.go#L71). 

So yes, the implicit expectation today is to use discovery, and pass in group and kind.

> Pinning a reference to a specific version has pros and cons:

One note: many single-kind references do some form of implicit pinning, by using the structured client libraries in things like client-go/listers. So there is some precedence for hard-pinning, although it's a bigger question of whether that's the practice we want to continue with.

agreed on pros and cons.

Reply:
One note here is that almost all references in Kubernetes core today are not generic references: they usually pare things down far enough that they do not need the full set of qualifiers {group,kind,version}.

If we want to recommend schema to match the common patterns today, I would imagine the guidance would be "include the fields that you need to help qualify the resource type". so include kind if the groups are clear, and do kind, group if you can divine the version, and all three if the reference is truly generic.

BTW let me know if there's some SIG that would be a good place to raise this conversation. There's a lot to unpack, and happy to do it all via GitHub / async, but happy to discuss in some in-person sync as well.

Reply:
group,resources are more precise for nearly every interaction than group,kinds.  A single group,kinds can (and sometimes do) map to multiple group,resources.  While some libraries don't quite handle this, we should encourage API models that remove unnecessary ambiguity.

Reply:
I think apiVersion should ONLY be specified when you need that specific version (e.g. need to delve into the type).  Using it beyond that is unnecessary coupling.

Reply:
I think David is right, but I think it's a little sad because:
1. The reference, especially if it refers to a field, is pointing to a field within the "kind" of the resource, not the resource itself.
2. It forces the entire type ecosystem to become discovery-aware, which I would assume most aren't (can we somehow address that with the openapi?)

Reply:
There might be two different types of references.

The one that access a field:
```golang
type FieldReference struct {
    Field Path // Path is to be defined
    APIVersion string
    Kind string // Or reference
    Name string
}
```

And the one that accesses an object:
```golang
type FieldReference struct {
    Group string
    Kind string // Or reference
    Name string
}
```

Reply:
Somewhat prior arts on that: https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/crds/
Also see: https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/namereference.go

Reply:
While there are many examples of depending upon Kind.  I would be cautious of recommending a type schema that intentionally allows a one to many mapping like group,kind trying to resolve a group,resource.

There are also examples of using the unambiguous group,resource to good effect and avoiding issues in consumption: https://github.com/openshift/api/blob/master/config/v1/types_cluster_operator.go#L79-L95 .  In many cases, these developed out of the difficulties in handling learned in the garbage collection controller.

Reply:
Thanks for all the feedback! Before diving into the specific recommendations, I'd like to clarify the original intent, as well as next steps.

My hope was to clarify the existing best practices around object references, and I was using common patterns from the *existing* CRDs to divine them.

But it sounds like people believe that the existing CRDs and the object references within them are not reflective of practices we'd like to use in the future (e.g. it sounds like group / resource is a preferred choice for a qualifier than apiVersion). 

If that is the case, I may move toward some larger document to try to capture these design choices, and kick off a discussion in the appropriate SIG meeting. Does that sound appropriate?

Reply:
> kick off a discussion in the appropriate SIG meeting. Does that sound appropriate?

Sure.  I suggest starting with the apimachinery mailing list.

Reply:
Note that `apiVersion` is usually both group AND version, for historical reasons group doesn't get its own field.

I also agree that resource is easier for clients as then they won't have to do a discovery step.

Reply:
One case in which we could consider permitting people to omit group for an arbitrary reference is if the group of the referenced object is the same as the group of the referencing object. However, on reflection I don't think we should permit this exception, I'm imagining references getting copy+pasted in yaml files and suddenly referring to the wrong thing.

I think the version is needed if and only if you are going to give a JSONPointer or something to identify a specific field.  Otherwise, you're assuming the reference resolver will just know what to do, which is also OK.

Reply:
> One case in which we could consider permitting people to omit group for an arbitrary reference is if the group of the referenced object is the same as the group of the referencing object. However, on reflection I don't think we should permit this exception, I'm imagining references getting copy+pasted in yaml files and suddenly referring to the wrong thing.

Agreed, makes sense. I think part of the challenge will be in standardizing the scenarios in which fields are omitted. For example, group can also be omitted in the case where an object reference is limited to a specific set of kinds, since the controller could just map the 'kind' (or whatever qualifier) to a map that populates the other missing qualifiers (version, group).

> I think the version is needed if and only if you are going to give a JSONPointer or something to identify a specific field. Otherwise, you're assuming the reference resolver will just know what to do, which is also OK.

I think that's the implicit assumption, but it would be good to qualify that there are specific things that the controller is expected to handle (e.g. selecting a specific version of a resource). I'll try to include these in the summary doc.

Reply:
> group can also be omitted in the case where an object reference is limited to a specific set of kinds

That is true, but how do we make that obvious to users? How do we prevent them from e.g. copy-pasting references that are valid in one place to another place where they are no longer valid?

Reply:
> How do we prevent them from e.g. copy-pasting references that are valid in one place to another place where they are no longer valid?

I think there is some risk, in addition to the fact that it is possible to, at some point, reference two kinds with the same name that belong to a different group.

In the name of making things easier to read than easier to write, I agree that we should incorporate group even if it's not strictly necessary.

> While there are many examples of depending upon Kind. I would be cautious of recommending a type schema that intentionally allows a one to many mapping like group,kind trying to resolve a group,resource.

@deads2k can you explain what scenario results in a (group, kind, version) resolving to multiple resources? From what I can tell CRDs are uniquely identified by either "kind" or "plural" fields.

Finally, I've [authored a doc](https://docs.google.com/document/d/1N2MbLsO0xLCa4PZkEM8g3Skt7XKv8C49_6yW1gBiMic/edit#heading=h.h81ukpnjbts9) that I believe incorporates all of the feedback. The only contentious choice I believe is kind vs resource.

Reply:
> @deads2k can you explain what scenario results in a (group, kind, version) resolving to multiple resources? From what I can tell CRDs are uniquely identified by either "kind" or "plural" fields.

A common case is `AdmissionReview.v1.admission.k8s.io`.  One easy way to get authentication for your admission webhook for free is it add it as an aggregated apiserver, then reference the kube-apiserver as the source.  This way every client can sent serviceaccount tokens automatically for authentication.  In some clusters, that kind can reference multiple resources.

Another example is `SubjectAccessReview.v1.authorization.k8s.io` for different kinds of webhook authorizers.  Also `TokenReview.v1.authorization.k8s.io`.

The same apimachinery capability is used to enable Scale.v1.autoscaling to work against multiple subresource endpoints.

We also used it for `Binding.v1.` against /binding and pods/binding.

Just because you don't often see examples, doesn't mean they don't exist.  Library code, APIs, and the architectural recommendations should reflect what the system is designed and used to do.  In this case, kind has ambiguous mappings to resources and URLs.  Using resource will eliminate this ambiguity.

Reply:
Thanks for the details! IIUC, the common use case of multiple resources sharing the same kind has to do with scoping the resource as a child of some other resource (e.g. top-level binding vs pod binding, deployment scale vs replicationcontroller scale)?

Reply:
Feel free to re-open this, but I've updated the guidance to use resource + group (and version as-needed) as suggested